### ISSUE TITLE
Panel issue under Linux

### ISSUE THREAD

**Issue by ivan-w** [SDL-Hercules-390/hyperion#289]
There seems to be a problem with the panel thread at startup under very specific conditions:
- The host is linux
- The host only has a single CPU/Core/thread
- The configuration has no additional file descriptors open other than stdin/stdout (no socket devices connected)
- hercules is being run as root (thus giving very high priority (-7) to the panel)

Tracing seems to show it is looping around the select() around line 1770 in panel.c:

https://github.com/SDL-Hercules-390/hyperion/blob/2b6b89923201b115c08093c5e0488c8ec13ff89c/panel.c#L1756-L1777

Input is very sluggish. every key takes a couple a seconds to show, and there is no real response to the command.

However, once the register panel is displayed (via ESC key) and turned back, everything works normal again.

With gdb, it seems the 'readset' is now different.

Investigating!

**Comment by Fish-Git** [comment]
`EINTR`?

**Comment by ivan-w** [comment]
Nah.. I checked that - it ain't it...

**Comment by ivan-w** [comment]
Good thing is I am able to reproduce the problem on my own test systems, so I can play, break and fiddle with everything now !

**Comment by ivan-w** [comment]
Ok - symptom : 
- Start hercules as root under a single processor VM...
- Basic hercules config, no funky devices (just 3270 and 3380s)
- Type "devlist"... it takes 10 seconds to show I actually types this... no answer
- Type the <ESC> key - gets me directoy to register display mode...
- Type the <ESC> key - I see the answer to the original "devlist" command and everything is responsive again. Immediate keyboard feedback and response to the commands.

Also the panel thread is completely hogging the CPU before you hit the <ESC> key.

And also if the configuration has devices that use TCP sockets, the problem doesn't show.

**Comment by Fish-Git** [comment]
Interesting... &nbsp; _(and weird!)_

Let us know what you find out. This is definitely a weird problem. I wonder why pressing the \<ESC\> key fixes things? Weird. Very weird...

**Comment by moshix** [comment]
Following my original reporting of this problem to Ivan, I have done some more tests. There are other weird symptoms:

If you use SDL current to try to IPL MVS TK4 with 2 CP defined, the MIPS counter will indicate 180-200MIPS activity with 5-6 SSIO per second, and not not manage to IPL the system at all. Not even the first nucleus message OS/VS2... etc shows up. Nothing.

**Comment by ivan-w** [comment]
Actually (for info) the problem even BEFORE IPLing anything....

**Comment by Juergen-Git** [comment]
> Following my original reporting of this problem to Ivan, I have done some more tests.
> There are other weird symptoms:
> If you use SDL current to try to IPL MVS TK4 with 2 CP defined, the MIPS counter will
> indicate 180-200MIPS activity with 5-6 SSIO per second, and not not manage to IPL the
> system at all. Not even the first nucleus message OS/VS2... etc shows up. Nothing.
>
Hi Moshe

As I'm currently (once again!) trying to come closer to releasing TK4- Update 09, I was wondering about your specific test scenario. Using vanilla TK4- Update 08 with the SDL-binaries update applied, I can't reproduce what you are describing. Of course I don't want to go into building the binaries for all the many platforms TK4- is supporting if there is a specific show stopper in the current code.

I'd appreciate any enlightening ;-)

Cheers
Jürgen

**Comment by moshix** [comment]
Juergen

Sure. I will try to explain as clearly as I possibly can. 

1st problem with SDL is when you run Hyperion on a 1 CPU (one thread only). You might say all modern CPUs have at least dual cores etc. But the cheapest instances at AWS, Google Cloud and Azure etc. are all single thread CPUs. 

When you have such a system and you configure TK4 with two CP, then the problem I described above occurs reliably on all common Linux distributions. Didn't try Windows.  

N.B.: It also sometimes occurs (on Centos 7 only so far) when TK4 has only one CP configured and the underlying LInux has one single CPU. 

The symptoms are very busy CP(s): with a few SIOs, and when you press ESC between the console log view and the register view, the responsiveness improves somewhat. 

IPL fails in all instances. 

2nd problem is that if you release TK4 Update9 with your own Hercules inside it, inevitably a lot of people will start to use that same Hercules for their other operating systems (because they want to avoid the hassle to build their own SDL). It is well known that current SDL will trigger IFI004/PROG004 and crash when

a) the disk subsystem is slow (for instance in cloud instances with spinning disks)
b) the VM CP is very busily IPLing other 64bit operating systems

Hope this helps

thanks

M

**Comment by Juergen-Git** [comment]
Thanks a lot, Moshe

for me that means I can continue as planned:

- Although I think that Hercules should throw a meaningful message when it chokes due to too little host resources being available, I don't think it should be a development goal to be able to run on the smallest cloud VMs you can possibly find. So, I'll ignore this panel problem for the TK4- use case.

By the way: The other mainframe emulator, Emily, poses clear and high prereqs on the underlying host system, and I think that's for good reason. We shouldn't go that far with Hercules but it is in the nature of multi threaded emulators that they become unusable when the underlying host provides too little parallelism.

- While your second problem may be valid in its own right, I think I can safely ignore it for TK4-. As you know, I see TK4- as an MVS 3.8j _appliance_, and thus I don't exactly care if people are using it for purposes it isn't designed for.

Otherwise the same is true as I said above: One simply shouldn't try to run operating systems that require a certain amount of parallelism and memory/disk resources on a host platform that cannot provide this. Of course, in this case too, Hercules should be nice enough to communicate this instead of letting the guest OS run into abends.

Cheers
Jürgen

**Comment by moshix** [comment]
Jürgen

Are you sure it's a resources problem and not a bug?  If I could do a shared "screen" session with you on Linux, you would see it's not a simple case of asking Linux or Hercules too much. 

It's a bonafide bug, and Ivan agrees with me. 

I should be able to IPL TK4 with one CP defined, on a one CPU Linux machine with 3.5GB of RAM, right?

I don't think this bug should be dismissed so easily because, for example myself, I run one CPU instances on Google cloud where Spinhawk IPLs TK4 just fine, and SDL doesn't. 

But of course you are the TK4 maintainer and you decide. 

Many thanks

M

**Comment by Juergen-Git** [comment]
Hi Moshe

no, I'm of course not sure whether it's a resource problem or a bug. If you and Ivan both think it's a bug, then it probably is. However, my statement wasn't about this question at all. I was simply about the fact that I don't see this problem as a showstopper for going on with TK4-, and that a good enough "solution" for me would be Hercules writing something like "low on host ressources, continue on your own risk", regardless, whether it is a bug or not.

Cheers
Jürgen

**Comment by Juergen-Git** [comment]
> I should be able to IPL TK4 with one CP defined, on a one CPU Linux machine with 3.5GB of RAM, right?
>
No, you can't state it like this. It totally depends on the CPU defined to or installed in your Linux machine. You can give it as much RAM as you like, but if the CPU doesn't provide enough parallelism and speed you'll have a problem.

**Comment by moshix** [comment]
ok, but if Spinhawk 3.13 on that very same machine IPLs TK4 without any issues and I get excellent performance on all my PL/1 and Cobol compile jobs, while SDL does not even IPL TK4 at all, then I think we are just closing our eyes to an obvious issue other than resources.

**Comment by Juergen-Git** [comment]
I agree with you in that all the Hyperions (not only the SDL one) are performance wise quite a bit behind Spinhawk, particularly in I/O intensive scenarios, which IPLs typically are. This already has been discussed in lengths and is mostly a consequence of longer instructions paths in the channel subsystem as compared to the original I/O implementation that still exists in Spinhawk. Well, but on the other hand a compliant channel subsystem is obviously a key requirement -- and the one we have in the Hyperions meets this requirement much closer than the old implementation. 

(just my 2¢)

**Comment by moshix** [comment]
Yes, I am aware that Hyperion are slower (due to the the threads for I/O). But I think slower means still being able to IPL TK4. 

The problem with SDL right is that it does not IPL, not even the nucleus of OS/VS2. Nothing. Not in one minute, not in on hour, not in hour day.

 If it just worked slow, then it should IPL but just take more time. However, failure to get MVS  to run, in my opinion is a serious bug which quite a few people will encounter with TK4 Update 9. 

After all the goodness you and the community have put into Update 9, it's a shame to have a distribution which will not IPL on the low-end cloud systems (which are exactly where you want to run a system like TK4 because the low-end systems are so cheap and come with more than enough resources to run a 24bit OS)

I respectfully suggest  we should investigate this bug and fix it before inclusion in TK4. I know Ivan is working on it.

**Comment by Juergen-Git** [comment]
Well, don't get me wrong: I never intended to prevent you and Ivan from investigating and resolving the issue. On the contrary!

I only said, I'll ignore the issue in the TK4- context. Given that TK4- Update 09 will run on stock SDL Hyperion, there will be absolutely no issue to replace the binaries once the issue is resolved.

**Comment by Fish-Git** [comment]
Out of curiosity, is the behavior any different (better or worse) when you run Hercules _not_ as root? (i.e. as a regular user?)

What about when you set [**`timerint`**](https://sdl-hercules-390.github.io/html/hercconf.html#TIMERINT) to, say, 500 (or even 5000) instead of letting it default to 50?

**Comment by moshix** [comment]
on Ubuntu 18.04, no different behavior root or sudo 

thanks

**Comment by Juergen-Git** [comment]
I just had a quick look into it, just for fun: It probably is simple race condition. The panel, running second highest priority, wins a race against someone else (the logger?), leading to a clobbered result from select(). A temporary fix can be found here:
https://polybox.ethz.ch/index.php/s/Zz8PNh30bxgNC2w 
Note: This only gives the "race partner" a better chance to win. I tested under Linux only, but the other platforms shouldn't be affected (at least not negatively). A definitive solution surely should probably identify the race partner and solve the conflict in a more friendly way.

**Comment by Juergen-Git** [comment]
Hi Fish

regarding your question on running as root and various TIMERINT settings: Yes, the problem occurs only when running as root. And I tried various settings of TIMERINT and PANRATE, all with no changes in the behaviour. Btw. TK4- runs with TIMERINT 10000 for many years know -- so this is good enough for MVS 3.8j and greatly reduced resource consumption when the OS is idle…

Cheers
Jürgen

**Comment by Juergen-Git** [comment]
… well, my patch from yesterday was a nice try, quick and dirty. But: It doesn't help in daemon mode, and from this can be concluded that the panel is only a casualty, i.e. _not_ part of the problem. So, basically, I made an intervention that works, but it takes place at the wrong location in the code. The working hypothesis still remains trying to fine a race condition and fix this then directly. When I'm at it now anyway I can follow on, unless Ivan or Fish volunteered desperately…

**Comment by Fish-Git** [comment]
> ... The working hypothesis still remains trying to find a race condition ...

It sure sounds like it. But then why doesn't it also occur on Windows? Why does it seem to only occur on Linux? Any ideas why that might be?

> ... unless Ivan or Fish volunteered desperately…

I would like to, but unfortunately my Linux debugging skills are extremely weak to non-existent. &nbsp; `:(`

**Comment by Fish-Git** [comment]
Maybe it has something to do with the `sysblk.panel_init` flag? Maybe we should change it into some type of condition variable under control of some type of lock? &nbsp; _&lt;shrug&gt;_

Or maybe we just need to change it to a completely separate flag BYTE flag instead? Right now it's a bit flag and we've had trouble in the past with compilers setting/testing bit flags? &nbsp; _&lt;shrug&gt;_

I'm just guessing here folks! I have _no idea_ what the frick is going on! &nbsp; `:(`

**Comment by Juergen-Git** [comment]
Thanks for the idea! Given that the panel as such can be ruled out (as even in daemon mode things are getting completely out of control), I think that we are seeing only secondary effects around the panel and its select() logic. When I make the panel work using my dirty trick from Wednesday, the system behaves in panel mode exactly the same as it does in daemon mode: At IPL the processor thread goes into a tight loop, but no instructions get executed by the guest…

I'm having a few ideas, but let me just do a few more experiments to get on the right track...

**Comment by Juergen-Git** [comment]
Weird, weird, weird: During some wild experimenting I just managed to accidentally bring up an old Hercules that lurked around somewhere on one of my test systems, build date in March 2018, classis Hyperion, i.e. _not_ an SDL version. Guess what: It exhibits the _exact_ same behaviour!

So, I'm starting to ask myself, when did we introduce this bug? Or shall we better go after my original proposal and simply call it a feature, the TLHRP (Too Little Host Resource Prevention) feature? Well, one of my bad jokes of course, I'll continue researching…

**Comment by Fish-Git** [comment]
> ... i.e. _not_ an SDL version. Guess what: It exhibits the _exact_ same behaviour!

Interesting!

> ... when did we introduce this bug?

Maybe it wasn't us? Maybe it's Linux itself?

As I recall, Linux's networking handling _was_ changed over the past year or three(*), and sockets (which _are_ networking objects!) _are_ involved in the problem we're having.

Does anyone have an "older" Linux they can test with? (To eliminate Linux networking handling itself from the list of suspects?) If the problem doesn't occur with older Linuxes (but does with newer ones), then we can blame Linux. Otherwise we'll just have to keep digging from the Hercules aspect.

Just a thought. &nbsp; `:)`
___
(*)  How significantly I don't know. But I seem to recall experiencing a frustrating period a few years back trying to deal with the different way it was handled between the older CentOS 6.x version I was using (and still am using!) at the time, and the newer (at that time) CentOS 7.x version. I think most of the frustration had to do with tuntap as I recall, but I also seem to recall it was the difference in the way networking itself was being handled between the two versions (old vs. new) that was at the root (heart) of the problem.

**Comment by Juergen-Git** [comment]
I'm currently also suspecting Linux kernel issues, be they simply behaviour changes or bugs doesn't exactly matter at this point.

The thing most probably has nothing to do with networking. The problem occurs whether networking devices are part of the game or not. Only when the panel problem (which as I said, probably is a casualty) is also present, the presence of networking devices may make a difference -- but that is then also purely accidental.

The real problem seems to be the priority handling (of the kernel), which also explains why the problems only occur when running as root (non root Hercules instances cannot modify thread priorities). I have now a temporary kludge that works: One or many host CPUs, in panel mode or in daemon mode, running as root or running as non root.

I consider it a kludge, as it only circumvents the suspected (kernel?) issue. Although it is not yet ready for general use, I think we should give it broad testing with the various Hercules uses cases that we know of, not only TK4-. The result of this testing would give an indication on whether it makes sense to produce a production version of the kludge and commit it for the time being, until someone comes up with a better idea (which would involve to deep dive into the pthread implementation on Linux).

I'll try to produce a patch for testing on Monday or Tuesday -- but it will still be quick and dirty.

**Comment by moshix** [comment]
many thanks!  I will be sure to test it thoroughly.

**Comment by wrljet** [comment]
Fish,

I have a Ubuntu 12.04 32-bit VM that I use regularly.  Is that old enough?
I'd test it out, if somebody tells me exactly what and how. 

Bill

**Comment by Fish-Git** [comment]
> I have a Ubuntu 12.04 32-bit VM that I use regularly. Is that old enough?

I have no idea. I cannot recall the exact date when the change(s) were made nor even _what_ the changes were (nor for that matter which distributions were affected or when). I'm not a Linux person.

> I'd test it out, if somebody tells me exactly what and how.

The only information we have is what Ivan and Jürgen posted earlier in the thread:

* https://github.com/SDL-Hercules-390/hyperion/issues/289#issuecomment-597978703
* https://github.com/SDL-Hercules-390/hyperion/issues/289#issuecomment-610882165

**Comment by Juergen-Git** [comment]
I just went back to an Ubuntu 14.04 32-bit, which also exhibits the problem. So, it would be nice to see your test on 12.04 32-bit, Bill. Just follow Ivan's instructions as mentioned above by Fish (the first link). If the panel problem is present, then your system has the problem (the panel problem suffices as an indicator).

**Comment by ivan-w** [comment]
The root problem is still in question. The 'select() call is acting as if there is data down the FD - but some action make it go back to normal behavior... Still dumbfounded here

**Comment by agadsby** [comment]
I’ve seen this type of behaviour in other systems. The issue is that Linux/Un*x select() has an inherent set of flaws that allow race conditions. The man page 

http://man7.org/linux/man-pages/man2/select.2.html

Explains some of the race conditions but there are others. This may also explain why Windows behaves differently. 

As a general guide the fds used in the select also need to be protected against being closed and reused by other threads whilst the select is active. If not you can get a select pop but the read at a later point in time is actually from a different source to the original fd. Image another thread closing the popping fd and the fd being reused in another open. 

Where is the select in question being called from?

Andy

**Comment by wrljet** [comment]
I tried it with the tk4-_v1.00_current and it seems to work fine with no unusual delays for me on Ubuntu 18.04 and 12.04 riding on a single CPU VMware VM.

Where do I find the beta TK4- Update 09?
Bill

**Comment by Juergen-Git** [comment]
There isn't yet an official beta of Update 09. However, to allow early testing of the binaries, I created a set of binaries for the various platforms that TK4- supports. They can be installed over an existing Update 08 installation. Please download from https://polybox.ethz.ch/index.php/s/sZk2wRMhsF2uwai/download and follow the README for installation.

**Comment by wrljet** [comment]
OK, with those SDL binaries applied, it hangs up on both Ubuntu 18.04 64-bit and 12.04 32-bit VMs.

On 12.04, 'top' shows CPU pegging at 95% in 'panel_display'.  The entire Linux is nearly unusable at that point.  I waited quite a long time, perhaps 5 minutes and nothing happened.
ESC to the other screen and back and things are back to normal.

So whatever is wrong was wrong in Ubuntu 12.04 as well.
Bill

**Comment by Juergen-Git** [comment]
Thanks, Bill, so as expected: Wherever this behaviour has its origin -- in Hercules or in Linux -- it is vvveeeerrryyy old :-). Weird, that up to now nobody ran into it.

@All, stay tuned for my temporary patch (kludge), tentatively available tomorrow.

Cheers
Jürgen

**Comment by wrljet** [comment]
Linux itself gets so sluggish with one CPU core.  Maybe not many people have been running with a single core CPU in recent times, things just went unnoticed.

Bill

**Comment by Fish-Git** [comment]
This thread, this GitHub Issue, this (Hercules?) bug, is VERY interesting, not to mention weird, perplexing, frustrating, annoying and many other various adjectives.

While I will certainly wait to see what Jürgen has come up with before seriously offering the following as a possible solution (due to its radical sledge-hammer approach), BUT...

One might pin the actual blame on Jan Jaeger, since, AFAIK, he was the one that designed our current file descriptor pipe design to Hercules message handling (panel thread, logger thread, etc). I absolutely recall (not the specific dates but definitely the conversation) my having voiced my opinion that it was a mistake (poor design) to use pipes to simply pass messages (panel/log messages) from one thread to another.

I felt _(and still feel!)_ a much better solution would have been a simple producer/consumer approach/design wherein messages are simply added to a queue (or queues) at the moment they are created and then removed from the queue(s) by those interested. Such a design is much simpler and more straightforward (not to mention more portable and less problematic IMO).

But alas, Jan said no. He argued that using pipes was a very commonly used technique for inter-thread communication and doing so was simpler and more straightforward than my approach. I told him at the time I felt it was a mistake.

That being said, I feel we should at least _consider_ a complete redesign of our current message handling as a _possible_ last-ditch approach to finally and forever resolving to this problem (depending on the outcome of Jürgen's currently unknown fix of course). That is to say, if Jürgen's fix ends up not working out (doesn't fix the problem) or seems too "klunky" (or kludgy) then we should at least _consider_ my approach to resolving this issue/problem finally and forever once and for all.

Comments?

**Comment by agadsby** [comment]
I've been playing with this. Using a pi with 4 cores and top to monitor I created a conf file as described and started hercules, no IPLm and I see the timer_thread using almost a whole core. The panel select() is popping as expected on the expired time interval. To me it seems the issue is related to what that timer thread is doing when an operating system is not IPLed - the main loop is very busy using a complete core, it does not do this when an real operating system is actually IPLed.

On a single core this would definitely make the system unresponsive. 

I'm now investigating why the timer_thread is behaving in this way.

**Comment by Juergen-Git** [comment]
All

… still evaluating. That thing is more complex than I expected ...

I now meandered for two days through various areas of the Hercules code and ended up with almost no result. Basically, my assumption now is, that everything I did during the past days in this matter, plus much of the work I did a year ago around libtelnet and console.c, plus the weird errors in comm3705, commadpt (async) and ctcadpt (CTCE), all have a common root cause: Sluggish piping between threads, particularly when I/O threads are part of the game (because the "slower" channel subsystem adds to the problem in this case). Thus, I think that Fish's proposal makes much sense, even if it comes with quite some risk and lots of work.

Without going into the dirty details of the past days, I can now say, that my "solution" probably isn't more than yet another workaround for the particular incarnation we are seeing here of this very old issue.

Having said this, my workaround could help for the time needed to pursue Fish's idea regarding message queues.

I'll now stop further experimenting and put the patch together as is, tomorrow. You'll see it's simple, almost trivial (just playing a bit with priorities), a typical indication of being a workaround only.

Cheers
Jürgen

**Comment by Juergen-Git** [comment]
Meant of course CTCT above, not CTCE...

**Comment by Juergen-Git** [comment]
So then, please find the kludge here:

https://polybox.ethz.ch/index.php/s/pCjHQknwA2q6J7r

You'll probably laugh and for me it's hard to believe too, that these minimal adaptions should make a difference at all. However, at least on my single socket/single core test VM they do. I'd be interested to see if they change the behaviour in other environments too. And, to re-iterate, we are talking about the weird panel behaviour in single threaded Linux systems, when running as root. When not running as root, or running on Windows, nothing should change at all.

If it really works in at least a few environments, then we can consider it being a workaround until we succeed finding the root cause or solving this and other issues using Fish's sledge hammer proposal...

**Comment by moshix** [comment]
Tried in my single CPU google cloud instance with ubuntu 14.04 and it solves the problem!

thanks

**Comment by ivan-w** [comment]
On a single CPU cloud instance, the issue may be hidden by the fact there are multiple actual underlying processors....

**Comment by ivan-w** [comment]
Moshe, I know you probably have a single CPU/CORE engine - which would be the best test case

**Comment by moshix** [comment]
How so? the kernel only sees and activates one CPU. 

Plus, with SDL I have the problem, with the patch applied, I don't...

**Comment by ivan-w** [comment]
Oh so someone found a solution to the problem ! Great !

**Comment by Juergen-Git** [comment]
Well, no, I didn't find a solution. I'd be glad if I found one, but the above patch is a circumvention in the best case -- or else it's just voodoo. There is no way out: Someone must find the root cause; otherwise that thing will arbitrarily bite again :-(.

**Comment by Fish-Git** [comment]
Thank you, Jürgen!

I agree that your "fix" is really just a workaround (circumvention) and that the actual root cause of the problem (no pun intended!) really needs to be (must be) identified by someone, somehow, or else this is just going to come back to bite us on the arse again at some later point.

Or... as I suggested, we should consider a complete redesign/rewrite of our inter-thread message handling.

OR... (option 2): maybe we just need to closely review our current thread priority handling? Windows's priority handling is quite different (obviously) from Linux's and/or POSIX's, and given that Jürgen's circumvention seems to work by simply tweaking the priorities of the threads involved, we might simply have a flaw in our POSIX thread priorities implementation? _&nbsp;&lt;shrug&gt;_

Just yet another thing to think about/consider.

So... given that Jürgen's patch does seem to resolve the problem for now (for the time being), should we maybe go ahead and commit it and close this issue? Or does his patch still need further testing/confirmation by more people?

**Comment by agadsby** [comment]
Please can we have a few days to investigate this further? I think I know why the select pops - it’s from the logger thread which doesn’t seem to get read by the panel loop. I have a test harness that I’m working on at the moment.

**Comment by Fish-Git** [comment]
Andrew Gadsby wrote:
> Fish wrote:
> > So... given that Jürgen's patch does seem to resolve the problem for now (for the time being), should we maybe go ahead and commit it and close this issue? Or does his patch still need further testing/confirmation by more people?
>
> Please can we have a few days to investigate this further?

_**NO!!  Absolutely not! Are you crazy?! We must install Jürgen's patch IMMEDIATELY!! The entire world will self destruct if we don't!!!**_

(heh)

_Of course_ we can wait a bit longer, Andrew. That's why I asked. &nbsp; `:)`

Especially given that it sounds like you're on to something. We would all be _very happy_ if you could manage to identify the actual cause of the problem, so _PLEASE_, take as long as you need!

**Comment by wrljet** [comment]
> 
> So then, please find the kludge here:
> 
> https://polybox.ethz.ch/index.php/s/pCjHQknwA2q6J7r
> 

Worked for me on Ubuntu 18.04 64-bit single CPU VM.
Very snappy response.

Bill

**Comment by agadsby** [comment]
Thanks Fish, I'm making headway - I suspect the full fix will be relatively simple. I'm currently unpicking how the logger thread gets data into the panel thread. There is a partial solution by removing the logger_thread and stdio FDs from the panel select - that fixes the panel CPU burn. However, it breaks the daemon mode by dropping the first few messages - this has pointed me at the root cause now and I'm working on that. I suspect the original CPU burn is caused by the logger writing to the logger_syslogfd FD but the panel doesn't clear that until you do a mode switch, hence that crazy CPU loop, with the high thread priority when root I suspect more data gets stuffed in leading to the select looping. 

Another couple of days, excluding this weekend, should get this nailed.

My partial fix to panel is below BUT this breaks daemon mode, by dropping some initial messages, at the moment - due to some interesting code in logmsg.

[ 7 line deletes and one line modified at the select() ]:

`--- panel.c     2020-04-16 14:14:18.659010696 +0100
***************
*** 1609,1611 ****
    int     rc;                           /* Return code               */
-   int     maxfd;                        /* Highest file descriptor   */
    fd_set  readset;                      /* Select file descriptors   */
--- 1609,1610 ----
***************
*** 1758,1765 ****
          FD_SET (keybfd, &readset);
-         FD_SET (logger_syslogfd[LOG_READ], &readset);
-         FD_SET (0, &readset);
-         if(keybfd > logger_syslogfd[LOG_READ])
-           maxfd = keybfd;
-         else
-           maxfd = logger_syslogfd[LOG_READ];

--- 1757,1758 ----
***************
*** 1769,1771 ****
          tv.tv_usec = (sysblk.panrate * 1000) % 1000000;
!         rc = select (maxfd + 1, &readset, NULL, NULL, &tv);
          if (rc < 0 )
--- 1762,1764 ----
          tv.tv_usec = (sysblk.panrate * 1000) % 1000000;
!         rc = select (keybfd + 1, &readset, NULL, NULL, &tv);
          if (rc < 0 )

**Comment by agadsby** [comment]
OK, I have a solution to the CPU burn problem. The immediate solution is to remove the use of the logger_syslogfd[LOG_READ] from the select() at line 1770 in panel.c line. The CPU burn is caused by the select() popping on data written to the logger pipe but not read within the loop. So changing the select line to:

`rc = select (keybfd + 1, &readset, NULL, NULL, &tv);
`
does the trick. The root cause of the issue is that data gets written to different places by logmsg.c depending on various fields in sysblk AND where stdin, stderr, stdout point to, so as threads start and exit sysblk fields change altering the route that messages go - if unlucky messages get sent down the logger pipe, triggering that select(), but are not then read within the select loop. This fits with Jurgen's suspicions that it was caused by thread timing.

There seems to be a lot of history in the logging code so I'd like to have a go at simplifying this - as a demonstration of how confusing the log flow is try this:

    hercules  >STDOUT   2>STDERR

You can see that the startup messages go to STDOUT and the shutdown ones to STDERR! This is a side affect of the code in logger.c starting line 382 
    /* ZZ FIXME: .... 
maybe that's a clue.

If the team agree I'd like to offer an immediate change to panel.c for the select() modification to fix, and close, the CPU burn issue AND continue to work on logger.c / logmsg.c / etc. to clean up the message flow - this will probably get rid of the logger pipe entirely.

**Comment by Fish-Git** [comment]
Andrew Gadsby wrote:

> OK, I have a solution to the CPU burn problem.

_**Excellent work**_, Andrew! &nbsp; _**THANK YOU!**_

> If the team agrees, I'd like to offer an immediate change to panel.c for the `select(`) modification to fix, and close, the CPU burn issue, _AND_ continue to work on logger.c / logmsg.c / etc. to clean up the message flow to ultimately get rid of the logger pipe entirely.

I can't speak for the rest of the team of course, but for myself, you certainly have _my_ vote!

And as far as your proposed temporary fix is concerned, you certainly have my vote on that as well. If no one else on the team objects, I propose that we implement Andrew's temporary workaround to close this issue while he continues to work on a more permanent solution.

Does everyone agree?

**Comment by Juergen-Git** [comment]
Thanks, Andrew! For me your proposal is fine, this pipe has hit us way too often...

**Comment by agadsby** [comment]
I've queued up change **Stop panel select watching logger pipe #1** for the initial fix. I couldn't link it to this issue, maybe I need to read the git manual.

**Comment by Fish-Git** [comment]
_(below comment copied from Issue #280 thread)_

Andrew Gadsby (@agadsby) wrote:

> Hi Jurgen and team.
>
> I’ve put the fix in for the select CPU burn issue. I believe that is sufficient to fix the reported issue.

The fix is in _your personal fork_ of SDL Hyperion, yes. I can see that from the helpful GitHub mention in issue #289. But your fix is still _not_ in the official SDL Hyperion repository yet. It's only in _your_ repository (fork), but not in _our_ repository. &nbsp; `:)`

I've never done it myself so I'm not one to speak, but I believe there's a way to submit/create a "pull request" requesting that _we_ (the SDL Hyperion repository) please do a "pull" _from **your**_ repository (fork) of the commit that you made that fixed the problem. Then we could Merge your pull request to get your fix into our repository and you'd get credit for it.

**_OR...._** &nbsp; The much _faster_ and _easier_ way (easier for both you and me) would be, since the fix is so simple, for me to simply directly commit your fix to our repository myself _(but with you as the "Author" of the fix of course, so that you get proper credit for it)_.

So I think that's what I'm going to do. &nbsp; `;-)`

Give me a few minutes and I'll commit your fix and close this issue _(i.e. Issue #289!)._

**Comment by Fish-Git** [comment]
Fixed by commit bef8eb0012583c33619d007ed18df749451413f1.

Issue closed.

Andrew Gadsby? (@agadsby)

Even though this immediate issue has now been closed, _PLEASE_ continue with your effort to fully address (fix) the logger issue and let us know when you have it ready. (No rush!)

Thanks!

**Comment by agadsby** [comment]
I believe some of the wider issues are caused by the write to the syslog pipe not checking the return length and delaying when the logger thread is behind in logmsg.c I have a fix for this that both checks the return length, looping if necessary, and also sets the logger pipe to none blocking. As commadpt.c has time sensitive things taking place it may explain the wider issues that Jurgen reported. I’ll finalise my proposal and get it submitted shortly.

**Comment by agadsby** [comment]
I have pulled together my changes to address several issues with the logger process and the message handling, see my internal commit diffs => [log changes.txt](https://github.com/SDL-Hercules-390/hyperion/files/4702196/log.changes.txt)

These changes do two things
i) Keep the logger alive for longer during  Hercules' shutdown thus getting many more messages out via the logger and draining the pipe correctly.
ii) Make the log message pipe none blocking to prevent the write_pipe hanging, this adds checks to write pipe to try and get the messages out in a reasonable time - previously it could hang for a long time if the logger was slow.

Messages still seem to come out of other routes but I believe these changes form a reasonable start. Please can someone else try these changes to see if I've broken anything before I ask Fish to commit them?

**Comment by Fish-Git** [comment]
> Please can someone else try these changes to see if I've broken anything before I ask Fish to commit them?

FWIW, your changes seem quite reasonable to me! Thanks, Andrew!

Per your request however, I will wait until others have first confirmed your changes before I commit them. &nbsp; _(But thanks again! Much appreciated, Andrew!)_

**Comment by agadsby** [comment]
Hold onto to that commit - I've broken something in the logger shutdown element so it deadlocks during tk4 shutdown.

**Comment by Fish-Git** [comment]
Thank you, Andrew! I will once again wait a few days(?) (longer?) until others have had a chance to test your latest patch.

**Comment by Fish-Git** [comment]
Andrew (@agadsby),

Is 25 days long enough? &nbsp; `;-)`

I would like to get your changes into Hercules.

**Comment by agadsby** [comment]
Hi Fish,

 I agree. At least once it's in it'll get a good thrashing. I cannot update remotely as I'm not authorised so how would you like me to proceed? Or are you happy to take the changes proposed above and commit them?

Andy

**Comment by Fish-Git** [comment]
> Or are you happy to take the changes proposed above and commit them?

Your changes look fine to me!

Fixed by commit 34f4353b6e132304250315f6bbf51320a347ec4b.

I gave you full credit of course (you were the author of the commit, I was just the committer).

I did make some minor cosmetic changes to your code however, which I hope was okay! (refer to above commit for details) If it's not okay, let me know and I'll do what I can to make it right.

Thanks again for your great work, Andy! It's contributions over the years from people just like you that has helped to make Hercules the fine product it is today. Your efforts are _greatly_ appreciated!
